Provide options.scriptExtensions for 2.x

supasate opened the pull request:
### Description
Provide scriptExtensions option for custom loading.
Due to require.extensions is not used by some framework (e.g. jest). So, this feature allows specifying file extensions supported by custom loaders.

#### Related issues
strongloop/loopback#3204

### Checklist

<!--
Please mark your choice with an "x" (i.e. [x], see
https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments)
-->

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

bajtos commented:
@supasate thank you for the pull request! Does @clarkorz's https://github.com/strongloop/loopback-boot/pull/240#discussion_r107839885 apply here too?

> The third param `onlyScriptsExportingFunction` cause `.js` files of models will not be loaded. 
> Line 142 in utils.js checking `typeof require.extensions['.js'] === 'function'` which will be false when using jest.

bajtos commented:
@slnode  ok to test

bajtos commented on lib/compiler.js:
```diff
@@ -38,6 +38,8 @@ module.exports = function compile(options) {
 
   var appRootDir = options.appRootDir = options.appRootDir || process.cwd();
   var env = options.env || process.env.NODE_ENV || 'development';
+  var scriptExtensions = options.scriptExtensions = options.scriptExtensions ||
```
Modifying input arguments (setting `options.scriptExtensions`) is a bad practice, we have been bitten by this in the past. Please treat `options` are readonly/frozen.

I see that you are accessing `options.scriptExtensions` at https://github.com/strongloop/loopback-boot/pull/242/files#diff-ee97c5091b89979aace94674818996baR722.

Here are two of many possible solutions how to keep `options` frozen and still have `scriptExtensions` set:

```js
// use prototypal inheritance
options = Object.create(options);

// shallow-copy properties
options = util._extend({}, options);
```

(A line like the above would go to L38-39 above, before we start processing the `options` object.)

bajtos commented on lib/compiler.js:
```diff
@@ -204,9 +209,10 @@ function tryReadDir() {
+    modelDefinitions, scriptExtensions) {
+  var registry = verifyModelDefinitions(rootDir, modelDefinitions,
+                  scriptExtensions) ||
+                  findModelDefinitions(rootDir, sources, scriptExtensions);
```
The indentation makes it difficult to distinguish function parameters from the second function call. I am proposing the following style:

```js
  var registry = verifyModelDefinitions(rootDir, modelDefinitions,
                                        scriptExtensions);

  if (!registry) {
    registry = findModelDefinitions(rootDir, sources, scriptExtensions);
  }
```

bajtos commented on lib/compiler.js:
```diff
@@ -38,6 +38,8 @@ module.exports = function compile(options) {
   var appRootDir = options.appRootDir = options.appRootDir || process.cwd();
   var env = options.env || process.env.NODE_ENV || 'development';
+  var scriptExtensions = options.scriptExtensions = options.scriptExtensions ||
+                                                    _.keys(require.extensions);
```
Considering that you are converting this  array back into an object down in `isPreferredExtension`, I think it's better to make the conversion right here.

```js
var scriptExtensions = options.scriptExtensions ?
  arrayToObject(scriptExtensions) : 
  require.extensions;
```

bajtos commented on lib/compiler.js:
```diff
@@ -644,19 +657,28 @@ function getExcludedExtensions() {
+
+function isPreferredExtension(filename, extensions) {
+  var includeExtensions = extensions ?
+    arrayToObject(extensions) : require.extensions;
```
Since `isPreferredExtension` is a private method, there is no need to keep support for the old one-arg API.

```js
function isPreferredExtension(filename, includeExtensions) {
  assert(!!includeExtensions, '"includeExtensions" argument is required');
  // the code below remains unchanged
  var ext = path.extname(filename);
  return (ext in includeExtensions) && !(ext in getExcludedExtensions());}
```

bajtos commented on test/fixtures/simple-app/boot/custom.customjs:
```diff
@@ -0,0 +1,8 @@
+  process.nextTick(function() {
+    process.bootFlags.push('customFinished');
+    callback();
+  });
```
This is unnecessarily complex. The thing to verify in your test is whether loopback-boot recognises a custom extension. There is no need to verify how boot script are loaded/executed again.

```js
'use strict';

module.exports = function(app, callback) {
  process.bootFlags.push('customjs');
  callback();
};
```

Same comment applies to `custom.customjs2`.

bajtos commented on test/executor.test.js:
```diff
@@ -299,6 +299,27 @@ describe('executor', function() {
       });
     });
 
+    it('should load scripts specified in options.scriptExtensions',
```
```js
it('searches boot file extensions specified in options.scriptExtensions')
```

See also http://loopback.io/doc/en/contrib/style-guide.html#test-naming

bajtos commented:
@supasate also please take a look at https://github.com/strongloop/loopback-boot/pull/240#discussion_r107839885 - please add a test reproducing the issue (I think the trick will be to use a model file with a custom extension that's not registered in `required.extensions`.)

bajtos commented:
Also please rebase your patch on top of the latest `2.x` branch, I fixed the failing Node 0.10 builds in #244.

supasate commented:
Hi @bajtos, I've rebased and fixed issues based on your comments. I made a sample repo at https://github.com/supasate/loopback-boot-with-jest. I'll squash later if it's alright.

supasate commented:
I also found another issue about file name resolving (https://github.com/strongloop/loopback/issues/2201) (not related to this PR but it will make `jest` failed to run).

bajtos commented on lib/compiler.js:
```diff
@@ -670,51 +699,52 @@ function fixFileExtension(filepath, files, onlyScriptsExportingFunction) {
 }
 
-function resolveAppScriptPath(rootDir, relativePath, resolveOptions) {
+function resolveAppScriptPath(rootDir, relativePath, resolveOptions,
```
Please keep `resolveOptions` as the last function arg. I would consider converting `scriptExtensions` into a property in `resolveOptions`, to keep the number of function arguments reasonably short.

supasate replied:
After refactoring, there is no need to use `scriptExtensions` in this function anymore.

bajtos commented on lib/compiler.js:
```diff
@@ -670,51 +699,52 @@ function fixFileExtension(filepath, files, onlyScriptsExportingFunction) {
+                                   scriptExtensions, normalization,
+                                   modelInstructions) {
   // load mixins from `options.mixins`
-  var sourceFiles = options.mixins || [];
```
What is the rationale for this change? I mean why to pass `options` properties as individual arguments? `buildAllMixinInstructions` is already accepting way too many arguments, we should look into ways how to make that list shorter, not longer...

supasate replied:
I tried to make code look consistent with [how `mixinDirs` and `mixinSources` are declared  and passed to the function.](https://github.com/supasate/loopback-boot/blob/8fef5f6fa8b0d2e37b1c38877805fbfeb656dbd6/lib/compiler.js#L112-L113).

Do you want to refactor those arguments by passing only `options` and set their default values in `buildAllMixinInstructions` instead?

supasate replied:
I refactored by passing only `options`.

bajtos commented on lib/compiler.js:
```diff
@@ -670,51 +699,52 @@ function fixFileExtension(filepath, files, onlyScriptsExportingFunction) {
-          (typeof require.extensions[otherFileExtension]) === 'function') {
-          results.push(otherFile);
-        }
+        // if (!onlyScriptsExportingFunction)
```
We don't keep commented-out code around, please remove.

If I understand this code correctly, the intention was:

 - return a file with any extension when `onlyScriptsExportingFunction` is not set
 - return a file with an extension that can be loaded by Node.js runtime (an extension that is registered in `require.extensions`) when `onlyScriptsExportingFunction` is set

We should preserve these two modes.

I am proposing to replace `onlyScriptsExportingFunction` arg with `scriptExtensions` and modify all function callers accordingly.

```js
// implementation here
if (!scriptExtensions || otherFileExtension in scriptExtensions)
  results.push(otherFile);

// usage - flag not set
fixFileExtension(filepath, files)

// usage -flag set
fixFileExtension(filepath, files, scriptExtensions)
```

supasate replied:
Oops! That's my mistake. I tested something and forgot to uncomment. I'll fix as you suggested.

bajtos commented:
@supasate thank you for the pull request, please take a look at the comment I left above. Also please rebase your patch on top of the latest `2.x` branch.

supasate commented:
About rebasing on top of the latest `2.x` branch, do you mean latest of `2.x` branch or `2.x-latest` branch? (I've rebased with `2.x` branch because it seems its last commit is more up-to-date than `2.x-latest`.)

bajtos commented:
Landed, thank you for the contribution!

bajtos commented:
Released in `loopback-boot@2.24.0` 🎉

supasate commented:
Thanks @bajtos ! 

I think you can review the 3.x branch also. The code is almost the same except that it is splitted into several files.
